Еще из 5го (и 6го)

GlowBlood opened the pull request.

GlowBlood committed:
Test 1

GlowBlood committed:
Lesson 2_Task1

GlowBlood committed:
Переделанный 2 Урок Таск 1

GlowBlood committed:
Опять исправления

GlowBlood committed:
Исправление про градусы

GlowBlood committed:
А вы знаете что тут у нас? ^_^

GlowBlood committed:
Исправила опечатку в легкой задаче

GlowBlood committed:
Вот так?

GlowBlood committed:
Некоторые задачи из Урока 3 Таск 1

GlowBlood committed:
Некоторые задачи из 4 урока

GlowBlood committed:
Исправления третьего урока

GlowBlood committed:
Исправления четвертого урока

Kavakl committed:
Попробовала исправить

GlowBlood committed:
Две сложные задачи третьего урока

Commit:
Парочку задач из 5го урока

Commit:
Парочку задач из 5го урока

Commit:
Исправленные пробелы в 3 уроке

Commit:
Исправления по стилю в задачах

Commit:
Испр опечатку

Commit:
Еще несколько задач из 5го, одна из которых не работает, но я буду доделывать

Commit:
Еще задачи из 5го

Commit:
Merge remote-tracking branch 'origin/master'

GlowBlood commented:
То ничего не отправлялось, то все сразу отправилось ¯\_(ツ)_/¯
Еще и лажа какая-то происходит

mglukhikh commented:
См. файл `Parse.kt`. Директивы импорта должны идти сразу после директивы `package`, нельзя вставлять между ними какие-либо определения -- `val MONTHS` в вашем случае. Переставьте ваш список месяцев ниже, и всё будет в порядке. Напоминаю, кстати, что перед созданием коммита следует проверить, проходят ли тесты в учебном проекте локально -- в вашем случае проект даже не компилируется.

Commit:
Исправила про список и импорт
Задачу, которая не компилируется, решила оставить пока, тк с другими не успеваю.

Commit:
Несколько задач из 6го урока (помню, что нужно исправить еще  из 5го)

mglukhikh commented:
Я посмотрю код немного позже, но обратите внимание на падающий тест в отчёте бота

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -1,6 +1,9 @@
+import lesson3.task1.maxDivisor
+val MONTHS = listOf("января", "февраля", "марта", "апреля", "мая", "июня", "июля", "августа", "сентября", "октября", "ноября", "декабря")
 
+val slist = listOf("0","1","2","3","4","5","6","7")
```
`"0123456789"`. По-моему, гораздо короче. А ещё можно добавлять к символу `'0'` цифру и получать соответствующий ей символ.

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,14 +57,20 @@ data class Circle(val center: Point, val radius: Double) {
      */
-    fun contains(p: Point): Boolean = TODO()
+    fun contains(p: Point): Boolean =
+    center.distance(p) <= radius
```
Форматирование поехало. Да и лучше было написать всё в одной строке, она не длинная

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +129,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
  */
-fun bestLongJump(jumps: String): Int = TODO()
+fun bestLongJump(jumps: String): Int {
+    val newJumps = jumps
```
Зачем эта переменная?

mglukhikh commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +129,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+       val newNewJumps = newJumps.split(" ").filter { it.toString().contains(Regex("""\d""")) }
+        var max = newNewJumps[0].toInt()
+        for (i in 0..newNewJumps.size - 2) {
+            if (newNewJumps[i + 1].toInt() > newNewJumps[i].toInt()) max = newNewJumps[i + 1].toInt()
```
Эта методика определения максимума в списке неверна в корне. А ещё можно было преобразовать всё к `List<Int>` с помощью `map` и потом использовать `list.max()`.

GlowBlood replied:
Не получается так сделать
![image](https://cloud.githubusercontent.com/assets/22037087/20636110/e8725442-b377-11e6-8003-728b55935989.png)

mglukhikh replied:
А для кого я на лекции рассказывал про Nullability в Котлине и как с ней бороться? Функция `list.max()` возвращает `null`, если список пуст, поэтому тип её результата `Int?`. Вам необходимо этот `null` на что-то заменить (см. текст этой задачи). Сделать это можно, например, с помощью Элвис-оператора `?:`.

mglukhikh reviewed (changes requested):
В принципе, всё довольно хорошо, замечания минимальны. Но `bestLongJump`, конечно, с ошибкой.

Commit:
Спасибо за замечания, постаралась исправить

GlowBlood committed:
Исправление задачи из 5го

GlowBlood committed:
Исправление задач из 6го

GlowBlood commented:
Извините, а можно получить какой-нибудь комментарий к моим изменениям?

GlowBlood commented:
А еще можно поинтересоваться, что Вы подразумеваете под словосочетанием "строковая нотация"??? Этого не написано нигде, видимо, нужно сообразить самостоятельно, но у меня не выходит.

mglukhikh commented:
"Возвращает строковую нотацию для клетки. В нотации, колонки обозначаются латинскими буквами от a до h, а ряды -- цифрами от 1 до 8.". 

```
a8 b8 c8 d8 e8 f8 g8 h8
...
a2 b2 c2 d2 e2 f2 g2 h2
a1 b1 c1 d1 e1 f1 g1 h1
```

mglukhikh commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -132,8 +132,10 @@ data class Line(val point: Point, val angle: Double) {
-    var newAngle = Math.atan((s.end.y - s.begin.y) / (s.end.x - s.begin.y))
+    var newAngle = Math.atan2(s.end.y - s.begin.y, s.end.x - s.begin.x)
+    if (newAngle == Math.PI || newAngle == -Math.PI) newAngle = 0.0
     if (s.begin.x == s.end.x) newAngle = Math.PI/2
```
По идее, эта проверка не нужна, `atan2` должен сделать всё сам.

mglukhikh reviewed (approved):
Я думаю, что ошибки вы исправили и PR можно закрыть. Так держать!

GlowBlood commented:
Понятно, спасибо

mglukhikh commented:
recheck all
